### PR TITLE
Hydrogen and Nitrogen fluid properties

### LINKED ISSUE #12410: FluidProperties: add H2 and N2
## Rationale

I have been using high-accuracy formulations for hydrogen and nitrogen for a while now, and have finally got some time this morning to clean them up and to add them to the fluid properties module for others to use.

## Description

The EOS for these fluids is similar to the CO2 fluid properties, as they are based on a Helmholtz free energy. So I have added a base class for these types of EOS's (HelmholtzFluidProperties), where all of the thermodynamic properties such as enthalpy, density, etc are calculated from the free energy.

This way, the userobjects for CO2, H2 and N2 only need to implement the free energy, and transport properties such as viscosity and thermal conductivity.

## Impact

New high-accuracy fluids that people can use.

### PR DESCRIPTION
**Pull request by cpgr**
I had to regold a number of tests that used CO2FluidProperties as the formulation uses a slightly more accurate calculation of specific gas constant now, which caused some very small diffs in the tests.

Closes #12410

### CODE REVIEW

**Comment by cpgr** [comment]
Not sure why the pthreads combined test is failing - that exact test passes when using `porous_flow-opt` but fails using `combined-opt`.

**Comment by andrsd** [comment]
I added `do not merge`, because this has a class name conflict with RELAP-7. we have an internal library implementing nitrogen fluid properties and it also provides `NitrogenFluidProperties` class.

**Comment by cpgr** [comment]
Looking closer at the test that failed, I can see that the results are slightly different from the gold file now. I wasn't aware that this test uses CO2, so this PR would slightly  change the density way down in the decimal places, which will cause a slight change in saturation. I've regolded which should fix it.

**Comment by cpgr** [comment]
I could always rename these if you want - N2FluidProperties etc.

File: modules/fluid_properties/src/userobjects/HelmholtzFluidProperties.C [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+  params.addClassDescription("Base class for Helmholtz free energy fluid EOS");
+  return params;
+}
+
+HelmholtzFluidProperties::HelmholtzFluidProperties(const InputParameters & parameters)
+  : SinglePhaseFluidProperties(parameters)
+{
+}
+
+HelmholtzFluidProperties::~HelmholtzFluidProperties() {}
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/HelmholtzFluidProperties.C]: In C++11, we do not have to create them empty d-tors any more.

File: modules/fluid_properties/src/userobjects/HydrogenFluidProperties.C [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+    _p_critical(1.315e6),
+    _T_critical(33.19),
+    _rho_molar_critical(15.508),
+    _rho_critical(1000.0 * _rho_molar_critical * _Mh2),
+    _p_triple(7.7e3),
+    _T_triple(13.952)
+{
+}
+
+HydrogenFluidProperties::~HydrogenFluidProperties() {}
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/HydrogenFluidProperties.C]: Empty d-tor -> remove.

File: modules/fluid_properties/src/userobjects/NitrogenFluidProperties.C [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+    _p_critical(3.3958e6),
+    _T_critical(126.192),
+    _rho_molar_critical(11.1839),
+    _rho_critical(313.3),
+    _p_triple(12.523e3),
+    _T_triple(63.151)
+{
+}
+
+NitrogenFluidProperties::~NitrogenFluidProperties() {}
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/NitrogenFluidProperties.C]: Empty d-tor -> remove.

File: unit/src/NitrogenFluidPropertiesTest.C [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+ * considered satisfactory.
+ */
+// TEST_F(NitrogenFluidPropertiesTest, partialDensity)
+// {
+//   const Real tol = 50.0 * REL_TOL_EXTERNAL_VALUE;
+//
+//   REL_TEST(_fp->partialDensity(373.15), 1182.8, tol);
+//   REL_TEST(_fp->partialDensity(473.35), 880.0, tol);
+//   REL_TEST(_fp->partialDensity(573.15), 593.8, tol);
+// }
```
Comment (andrsd) [inline unit/src/NitrogenFluidPropertiesTest.C]: Was this supposed to be uncommented for the final version?

**Review by andrsd — CHANGES_REQUESTED** [review]

File: unit/src/NitrogenFluidPropertiesTest.C (hunk 4 above)
Comment (cpgr) [reply unit/src/NitrogenFluidPropertiesTest.C]: Bad copy and paste. It shouldn't be here

**Review by andrsd — APPROVED** [review]

**Pull request merged**